src/vcpkg/registries.cpp

```diff
@@ -96,11 +100,48 @@ namespace
             });
         }
 
+        struct VTPResult
+        {
+            fs::path path;
+            bool stale;
+        };
+
+        VTPResult get_stale_versions_tree_path(const VcpkgPaths& paths) const
+        {
+            auto e = get_lock_entry(paths);
+            if (!e.stale())
+            {
+                return {get_versions_tree_path(paths), false};
+            }
+            if (!m_stale_versions_tree.has_value())
+            {
+                auto maybe_tree =
+                    paths.git_find_object_id_for_remote_registry_path(e.value(), registry_versions_dir_name);
+                if (!maybe_tree)
+                {
+                    // This could be caused by git gc or otherwise -- fall back to full fetch
+                    return {get_versions_tree_path(paths), false};
+                }
+                auto maybe_path = paths.git_checkout_object_from_remote_registry(*maybe_tree.get());
+                if (!maybe_path)
+                {
+                    // This could be caused by git gc or otherwise -- fall back to full fetch
+                    return {get_versions_tree_path(paths), false};
+                }
+                m_stale_versions_tree = std::move(*maybe_path.get());
+            }
+            return {*m_stale_versions_tree.get(), true};
+        }
+
         std::string m_repo;
-        DelayedInit<std::string> m_commit; // TODO: eventually this should end up in the vcpkg-lock.json file
-        DelayedInit<fs::path> m_versions_tree;
         std::string m_baseline_identifier;
+        DelayedInit<LockFile::Entry> m_lock_entry;
+        mutable Optional<fs::path> m_stale_versions_tree;
+        DelayedInit<fs::path> m_versions_tree;
         DelayedInit<Baseline> m_baseline;
+
+        // Shared by all child GitRegistryEntry's
+        mutable const VcpkgPaths* m_paths = nullptr;
```

Comment:
Note to self: rip this out later. (also please make comments TODO-ing that, since we should keep a record of what we need to do)